src/librustc_trans/trans/consts.rs

```diff
@@ -591,6 +591,21 @@ fn const_expr_unadjusted<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
             let is_float = ty.is_fp();
             let signed = ty.is_signed();
 
+            if ty.is_bool() && (b.node == hir::BiOr || b.node == hir::BiAnd) {
+              match eval_const_expr_partial(cx.tcx(),
```

Comment:
yes you're right, see also @oli-obk 's comment. I would need to fix this better somehow, but don't yet know a good way.